control-plane/csi-driver/src/bin/controller/controller.rs

```diff
@@ -795,15 +859,73 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
         let snap_uuid = Uuid::parse_str(&snapshot_uuid_str).map_err(|_e| {
             Status::invalid_argument(format!("Malformed snapshot ID: {}", request.name))
         })?;
+        let create_params = CreateSnapshotParams::try_from(&request.parameters)?;
+
+        // Get the volume object. Extract the app node endpoint if the quiesce is requested.
+        let volume = RestApiClient::get_client().get_volume(&volume_uuid).await?;
+        let app_node_endpoint_info = match get_volume_app_node(&volume) {
+            // Volume is not published, so no need to quiesce.
+            None => None,
+            Some(info) => match create_params.queisce().clone() {
+                None | Some(QuiesceCandidate::Filesystem) => {
+                    // If quiesce is requested, get the app node endpoint. Request would fail to
+                    // proceed if app node endpoint is not retieved.
+                    let app_node = RestApiClient::get_client().get_app_node(&info).await?;
+                    Some(app_node.spec.endpoint)
+                }
+                Some(QuiesceCandidate::None) => None,
+            },
+        };
 
-        let snapshot = RestApiClient::get_client()
-            .create_volume_snapshot(&volume_uuid, &snap_uuid)
+        let snapshot_creation_result = match RestApiClient::get_client()
+            .get_volumes_snapshot(&snap_uuid)
             .await
-            .map_err(|error| match error {
-                ApiClientError::ResourceExhausted(reason) => Status::resource_exhausted(reason),
-                ApiClientError::PreconditionFailed(reason) => Status::resource_exhausted(reason),
-                error => error.into(),
-            })?;
+        {
+            // If snapshot is already created in previous attemps fetch and return it.
+            Ok(snapshot) => Ok(snapshot),
+            Err(ApiClientError::ResourceNotExists(_)) => {
+                if let Some(app_node_endpoint) = app_node_endpoint_info.clone() {
+                    issue_fs_freeze(app_node_endpoint, volume_uuid.to_string()).await?;
+                }
+
+                // Create the snapshot.
+                RestApiClient::get_client()
+                    .create_volume_snapshot(&volume_uuid, &snap_uuid)
+                    .await
+                    .map_err(|error| match error {
+                        ApiClientError::ResourceExhausted(reason) => {
+                            Status::resource_exhausted(reason)
+                        }
+                        ApiClientError::PreconditionFailed(reason) => {
+                            Status::resource_exhausted(reason)
```

Comment:
Why do this convert to resource exhausted?

Reply:
This is pre-existing bit of code to handle the snapshot-controller. We cannot return precondition failed on snapchat failure as that would not allow for deletion.